### ISSUE TITLE
Support very large test results

### ISSUE THREAD

**Issue by ben-manes** [EnricoMi/publish-unit-test-result-action#231]
My [project](https://github.com/ben-manes/caffeine) produces millions of test executions over a 2 hour build. I was hoping to use this action for a summary of the results, which would be a lightweight as just the counts at the suite level. Unfortunately the action [failed](https://github.com/ben-manes/caffeine/runs/5437723729?check_suite_focus=true) with an out of memory error (exit code 137). Can the results be processed in a streaming fashion rather than read fully into memory, which avoids having a limit?

A possible workaround that I might try is to delete all the test methods prior to processing, e.g.
```sh
xml ed -L -d "/testsuite/*" test.xml
```

**Comment by EnricoMi** [comment]
This are 6.31 GB of test results, wow! Let me see what I can do ...

**Comment by ben-manes** [comment]
haha, yes. I tried it on a local download and the cli tool crashed as it's not streaming either.

```
$ find . -type f -exec xml ed -L -d "/testsuite/*" {} \;
./TEST-com.github.benmanes.caffeine.cache.ExpirationTest.xml:124476.115: internal error: Huge input lookup
ache.testing.RemovalListeners$RejectingRemovalListener.onRemoval(RemovalListener
```

**Comment by EnricoMi** [comment]
Since you are running with `check_run_annotations: none`, you can safely remove all those `<testcase>` nodes under the root `<testsuite>`. The action takes the statistics from the `<testsuite>` and only needs the individual `<testcase>` nodes for annotations.

**Comment by ben-manes** [comment]
yes, exactly. I need to find the right incantation of a shell command to do that. Something like this should work, though not quite right yet: `find . -name "*.xml" -exec sh -c 'grep testsuite {} > {}' \;`

**Comment by ben-manes** [comment]
I think will do it as the preprocessing step.
```sh
find . -type f -name "*.xml" -exec sh -c 'grep testsuite {} > {}.out && mv {}.out {}' \;
```

closing and off to give it a try.

**Comment by EnricoMi** [comment]
Thanks for reporting, I am curious about any workaround that works for you.

**Comment by ben-manes** [comment]
kicked off a [job](https://github.com/ben-manes/caffeine/actions/runs/1942277927), but it will take ~2 hours for the full run.

**Comment by EnricoMi** [comment]
Can you test the action from this branch:

    uses: EnricoMi/publish-unit-test-result-action/composite@branch-xml-ignore-testcases

That should not require you modifying the XML files.

I may make this available through a new option. This branch is just a PoC.

**Comment by ben-manes** [comment]
Sure, here's that [job](https://github.com/ben-manes/caffeine/actions/runs/1942372569)

**Comment by EnricoMi** [comment]
The first job looks ok, the parsing errors will go away when you point `files` to test result xml files only:

     files: '**/*.xml'

The reported `0` tests will go away when you rerun the second job (assuming it has the same output as the first job, hasn't finished yet). There is a fix in the branch that did not get picked up by the second job as I pushed it after your workflow started.

**Comment by ben-manes** [comment]
I kicked off another [job](https://github.com/ben-manes/caffeine/actions/runs/1942580017) with that fix. 🙁

**Comment by EnricoMi** [comment]
And, as a side note: with matrix strategy, you have to give the jobs individual `check_name` (include `${{ matrix.java }}`) so that you get two results, otherwise the two results will overwrite each other.

**Comment by ben-manes** [comment]
right, I intend to use an artifact and separate `workflow_run` job, but that only runs if on the master branch. So I inlined these for debugging.

**Comment by ben-manes** [comment]
Looks like it [attached](https://github.com/ben-manes/caffeine/runs/5441188131) to the examples workflow, oddly enough, but still got a nice summary.

<img width="459" alt="Screen Shot 2022-03-06 at 3 18 50 PM" src="https://user-images.githubusercontent.com/378614/156946627-5b7ac762-d80b-450e-90c0-1b0b0c35f013.png">

**Comment by EnricoMi** [comment]
Your browser is running on macOS, right? On Ubuntu, this renders nicer:
![grafik](https://user-images.githubusercontent.com/44700269/156946962-edb3f6ea-b914-4689-b4d7-d66e0b164a20.png)
Amazing how different Github looks across OS.

**Comment by ben-manes** [comment]
Yep, that looks much nicer. Odd as probably both on chrome?

**Comment by EnricoMi** [comment]
Here is Ubuntu Chrome:
![grafik](https://user-images.githubusercontent.com/44700269/156947087-902a231c-f2b5-42b6-a3a8-8ca221fccd01.png)

**Comment by EnricoMi** [comment]
OK, so the branch works equally well, no need to preprocess the files then.

**Comment by EnricoMi** [comment]
This has been released as v1.31, available via v1.

**Comment by ben-manes** [comment]
Thanks! I [pushed](https://github.com/ben-manes/caffeine/runs/5454777521) the switch over to this new config option.

**Comment by EnricoMi** [comment]
Thanks for testing, I am watching that job.

**Comment by EnricoMi** [comment]
Yay, this worked:
![image](https://user-images.githubusercontent.com/44700269/157131190-35a8b797-e174-40b2-b711-a8a01e745506.png)

Next run, the lowest row will disappear, only leaving the test row.

8.5m tests, this is definitively top 1 in this action's high score.

**Comment by ben-manes** [comment]
This is great! I had to estimate counts before since it is so huge. 🙂 

For this project the complexity is very high and beyond what I can mentally track. I decided to brute force testing was a good safety net, even if sometimes excessive. The ~2000 tests methods are parameterized to run for every valid configuration (a cartesian product of the specification constraints), which catches a lot of simple mistakes during development that happen only in some cases. One of those settings is the backing implementation, where Google's Guava is a reference to assert similar behavior. I also ported unit test suite from the jdk, eclipse, apache, and google to fill in gaps due to lack of imagination. I then throw in many static analyzers, etc in hopes of catching whatever remains. Adding features or tests for bug fixes inches those numbers further. All of that is to say that the project is too complex for my meager brain so I copped out and burn cpu cycles instead 😄
